File: screener/backtester/pine.py
"""Safe Pine-like expression parser and evaluator.

Parses a small subset of Pine Script into an AST and evaluates it against a
pandas DataFrame of OHLCV bars. Does NOT use Python ``eval``.

Supported:
  series: open, high, low, close, volume, adj_close
  literals: int / float
  arithmetic: + - * /
  comparison: > >= < <= == !=
  boolean:    and, or, not
  functions:  sma(s, n), ema(s, n), rsi(s, n), highest(s, n), lowest(s, n),
              atr(n), crossover(a, b), crossunder(a, b)

Causality guarantee: every rolling / shift operation is left-aligned, so the
value at bar ``i`` depends only on bars ``<= i``. crossover/crossunder use a
single ``.shift(1)`` and so compare only bars ``i`` and ``i-1``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union

import numpy as np
import pandas as pd


class PineError(Exception):
    """Base class for Pine expression errors."""


class PineSyntaxError(PineError):
    """Raised when the expression fails to tokenize or parse."""


class PineNameError(PineError):
    """Raised when an unknown identifier or function is referenced."""


SERIES_NAMES = {"open", "high", "low", "close", "volume", "adj_close"}
ROLLING_FUNCS = {"sma", "ema", "rsi", "highest", "lowest"}
FUNC_NAMES = ROLLING_FUNCS | {"atr", "crossover", "crossunder"}
BOOL_KEYWORDS = {"and", "or", "not", "true", "false"}


# ── AST ──────────────────────────────────────────────────────────────


@dataclass
class Num:
    value: float


@dataclass
class Name:
    name: str


@dataclass
class UnaryOp:
    op: str  # '-' or '+'
    operand: "Node"


@dataclass
class Not:
    operand: "Node"


@dataclass
class BinOp:
    op: str  # '+', '-', '*', '/'
    left: "Node"
    right: "Node"


@dataclass
class Compare:
    op: str  # '>', '>=', '<', '<=', '==', '!='
    left: "Node"
    right: "Node"


@dataclass
class BoolOp:
    op: str  # 'and' or 'or'
    left: "Node"
    right: "Node"


@dataclass
class Call:
    name: str
    args: list
    col: int  # column in source (for error messages)


Node = Union[Num, Name, UnaryOp, Not, BinOp, Compare, BoolOp, Call]


# ── tokenizer ────────────────────────────────────────────────────────


_TWO_CHAR = {">=", "<=", "==", "!="}
_SINGLE_PUNCT = set("+-*/()><,")


@dataclass
class Token:
    kind: str  # 'num', 'name', 'op', 'lp', 'rp', 'comma', 'end'
    value: str
    col: int


def _tokenize(expr: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    n = len(expr)
    while i < n:
        ch = expr[i]
        if ch.isspace():
            i += 1
            continue
        if ch.isdigit() or (ch == "." and i + 1 < n and expr[i + 1].isdigit()):
            start = i
            saw_dot = ch == "."
            i += 1
            while i < n and (expr[i].isdigit() or (expr[i] == "." and not saw_dot)):
                if expr[i] == ".":
                    saw_dot = True
                i += 1
            tokens.append(Token("num", expr[start:i], start))
            continue
        if ch.isalpha() or ch == "_":
            start = i
            i += 1
            while i < n and (expr[i].isalnum() or expr[i] == "_"):
                i += 1
            tokens.append(Token("name", expr[start:i], start))
            continue
        two = expr[i : i + 2]
        if two in _TWO_CHAR:
            tokens.append(Token("op", two, i))
            i += 2
            continue
        if ch == "(":
            tokens.append(Token("lp", ch, i))
            i += 1
            continue
        if ch == ")":
            tokens.append(Token("rp", ch, i))
            i += 1
            continue
        if ch == ",":
            tokens.append(Token("comma", ch, i))
            i += 1
            continue
        if ch in _SINGLE_PUNCT:
            tokens.append(Token("op", ch, i))
            i += 1
            continue
        raise PineSyntaxError(f"Unexpected character {ch!r} at column {i}")
    tokens.append(Token("end", "", n))
    return tokens


# ── parser (recursive descent with precedence climbing) ─────────────


class _Parser:
    def __init__(self, tokens: list[Token]) -> None:
        self.tokens = tokens
        self.pos = 0

    def peek(self) -> Token:
        return self.tokens[self.pos]

    def consume(self) -> Token:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def expect(self, kind: str, value: str | None = None) -> Token:
        tok = self.peek()
        if tok.kind != kind or (value is not None and tok.value != value):
            expected = value or kind
            raise PineSyntaxError(
                f"Expected {expected!r} at column {tok.col}, got {tok.value!r}"
            )
        return self.consume()

    # entry point
    def parse(self) -> Node:
        node = self.parse_or()
        if self.peek().kind != "end":
            tok = self.peek()
            raise PineSyntaxError(f"Unexpected {tok.value!r} at column {tok.col}")
        return node

    def parse_or(self) -> Node:
        left = self.parse_and()
        while self.peek().kind == "name" and self.peek().value == "or":
            self.consume()
            right = self.parse_and()
            left = BoolOp("or", left, right)
        return left

    def parse_and(self) -> Node:
        left = self.parse_not()
        while self.peek().kind == "name" and self.peek().value == "and":
            self.consume()
            right = self.parse_not()
            left = BoolOp("and", left, right)
        return left

    def parse_not(self) -> Node:
        if self.peek().kind == "name" and self.peek().value == "not":
            self.consume()
            return Not(self.parse_not())
        return self.parse_compare()

    def parse_compare(self) -> Node:
        left = self.parse_add()
        tok = self.peek()
        if tok.kind == "op" and tok.value in {">", ">=", "<", "<=", "==", "!="}:
            self.consume()
            right = self.parse_add()
            return Compare(tok.value, left, right)
        return left

    def parse_add(self) -> Node:
        left = self.parse_mul()
        while self.peek().kind == "op" and self.peek().value in {"+", "-"}:
            op = self.consume().value
            right = self.parse_mul()
            left = BinOp(op, left, right)
        return left

    def parse_mul(self) -> Node:
        left = self.parse_unary()
        while self.peek().kind == "op" and self.peek().value in {"*", "/"}:
            op = self.consume().value
            right = self.parse_unary()
            left = BinOp(op, left, right)
        return left

    def parse_unary(self) -> Node:
        if self.peek().kind == "op" and self.peek().value in {"+", "-"}:
            op = self.consume().value
            return UnaryOp(op, self.parse_unary())
        return self.parse_primary()

    def parse_primary(self) -> Node:
        tok = self.peek()
        if tok.kind == "num":
            self.consume()
            return Num(float(tok.value))
        if tok.kind == "lp":
            self.consume()
            node = self.parse_or()
            self.expect("rp")
            return node
        if tok.kind == "name":
            self.consume()
            if self.peek().kind == "lp":
                self.consume()
                args: list[Node] = []
                if self.peek().kind != "rp":
                    args.append(self.parse_or())
                    while self.peek().kind == "comma":
                        self.consume()
                        args.append(self.parse_or())
                self.expect("rp")
                return Call(tok.value, args, tok.col)
            if tok.value in {"true", "false"}:
                return Num(1.0 if tok.value == "true" else 0.0)
            return Name(tok.value)
        raise PineSyntaxError(f"Unexpected token {tok.value!r} at column {tok.col}")


def parse(expr: str) -> Node:
    if not expr or not expr.strip():
        raise PineSyntaxError("Empty expression")
    return _Parser(_tokenize(expr)).parse()


# ── evaluator ────────────────────────────────────────────────────────


def _as_series(value, index: pd.Index) -> pd.Series:
    if isinstance(value, pd.Series):
        return value
    return pd.Series(float(value), index=index)


def _require_int_literal(node: Node, func: str, arg: str) -> int:
    if not isinstance(node, Num):
        raise PineSyntaxError(f"{func}() argument {arg!r} must be an integer literal")
    n = int(node.value)
    if n != node.value or n <= 0:
        raise PineSyntaxError(
            f"{func}() argument {arg!r} must be a positive integer, got {node.value}"
        )
    return n


def _rsi(source: pd.Series, length: int) -> pd.Series:
    delta = source.diff()
    gains = delta.clip(lower=0.0)
    losses = -delta.clip(upper=0.0)
    # Wilder smoothing: EMA with alpha = 1/length
    avg_gain = gains.ewm(alpha=1.0 / length, adjust=False, min_periods=length).mean()
    avg_loss = losses.ewm(alpha=1.0 / length, adjust=False, min_periods=length).mean()
    rs = avg_gain / avg_loss.replace(0, np.nan)
    rsi = 100.0 - (100.0 / (1.0 + rs))
    # if avg_loss is 0 and avg_gain > 0, RSI = 100
    rsi = rsi.where(~((avg_loss == 0) & (avg_gain > 0)), 100.0)
    return rsi


def _atr(bars: pd.DataFrame, length: int) -> pd.Series:
    high = bars["high"]
    low = bars["low"]
    prev_close = bars["close"].shift(1)
    tr = pd.concat(
        [
            (high - low).abs(),
            (high - prev_close).abs(),
            (low - prev_close).abs(),
        ],
        axis=1,
    ).max(axis=1)
    return tr.ewm(alpha=1.0 / length, adjust=False, min_periods=length).mean()


def _crossover(a: pd.Series, b: pd.Series) -> pd.Series:
    a_now = a
    b_now = b
    a_prev = a.shift(1)
    b_prev = b.shift(1)
    cond = (a_now > b_now) & (a_prev <= b_prev)
    return cond.fillna(False)


def _crossunder(a: pd.Series, b: pd.Series) -> pd.Series:
    a_now = a
    b_now = b
    a_prev = a.shift(1)
    b_prev = b.shift(1)
    cond = (a_now < b_now) & (a_prev >= b_prev)
    return cond.fillna(False)


def _series_from_name(name: str, bars: pd.DataFrame) -> pd.Series:
    if name == "adj_close":
        # with auto_adjust=True, close IS adjusted; alias them
        if "adj_close" in bars.columns:
            return bars["adj_close"].astype(float)
        return bars["close"].astype(float)
    if name in SERIES_NAMES:
        if name not in bars.columns:
            raise PineNameError(f"Series {name!r} not available in bars DataFrame")
        return bars[name].astype(float)
    if name in bars.columns:
        return pd.to_numeric(bars[name], errors="coerce").astype(float)
    raise PineNameError(f"Unknown identifier: {name!r}")


def _eval(node: Node, bars: pd.DataFrame):
    if isinstance(node, Num):
        return float(node.value)
    if isinstance(node, Name):
        return _series_from_name(node.name, bars)
    if isinstance(node, UnaryOp):
        val = _eval(node.operand, bars)
        if node.op == "-":
            return -val
        return val
    if isinstance(node, Not):
        val = _as_series(_eval(node.operand, bars), bars.index).astype(bool)
        return ~val
    if isinstance(node, BinOp):
        left = _eval(node.left, bars)
        right = _eval(node.right, bars)
        if node.op == "+":
            return left + right
        if node.op == "-":
            return left - right
        if node.op == "*":
            return left * right
        if node.op == "/":
            return left / right
        raise PineSyntaxError(f"Unknown operator: {node.op!r}")
    if isinstance(node, Compare):
        left = _eval(node.left, bars)
        right = _eval(node.right, bars)
        ops = {
            ">": lambda a, b: a > b,
            ">=": lambda a, b: a >= b,
            "<": lambda a, b: a < b,
            "<=": lambda a, b: a <= b,
            "==": lambda a, b: a == b,
            "!=": lambda a, b: a != b,
        }
        out = ops[node.op](left, right)
        if isinstance(out, pd.Series):
            return out.fillna(False)
        return bool(out)
    if isinstance(node, BoolOp):
        left = _as_series(_eval(node.left, bars), bars.index).astype(bool)
        right = _as_series(_eval(node.right, bars), bars.index).astype(bool)
        if node.op == "and":
            return left & right
        return left | right
    if isinstance(node, Call):
        return _eval_call(node, bars)
    raise PineSyntaxError(f"Unknown AST node: {type(node).__name__}")


def _eval_call(node: Call, bars: pd.DataFrame):
    name = node.name
    if name not in FUNC_NAMES:
        raise PineNameError(f"Unknown function: {name!r} at column {node.col}")
    if name in ROLLING_FUNCS:
        if len(node.args) != 2:
            raise PineSyntaxError(
                f"{name}() takes 2 arguments (source, length), got {len(node.args)}"
            )
        source_val = _eval(node.args[0], bars)
        if not isinstance(source_val, pd.Series):
            source_val = _as_series(source_val, bars.index)
        length = _require_int_literal(node.args[1], name, "length")
        if name == "sma":
            return source_val.rolling(length, min_periods=length).mean()
        if name == "ema":
            return source_val.ewm(span=length, adjust=False, min_periods=length).mean()
        if name == "rsi":
            return _rsi(source_val, length)
        if name == "highest":
            return source_val.rolling(length, min_periods=length).max()
        if name == "lowest":
            return source_val.rolling(length, min_periods=length).min()
    if name == "atr":
        if len(node.args) != 1:
            raise PineSyntaxError(
                f"atr() takes 1 argument (length), got {len(node.args)}"
            )
        length = _require_int_literal(node.args[0], "atr", "length")
        return _atr(bars, length)
    if name in {"crossover", "crossunder"}:
        if len(node.args) != 2:
            raise PineSyntaxError(f"{name}() takes 2 arguments, got {len(node.args)}")
        a = _as_series(_eval(node.args[0], bars), bars.index)
        b = _as_series(_eval(node.args[1], bars), bars.index)
        return _crossover(a, b) if name == "crossover" else _crossunder(a, b)
    raise PineNameError(f"Unknown function: {name!r}")


def evaluate(node: Node, bars: pd.DataFrame) -> pd.Series:
    """Evaluate ``node`` against ``bars`` and return a Series aligned to bars.index.

    Booleans are returned as a bool Series; numeric results are returned as a
    float Series. Pure-scalar results are broadcast across ``bars.index``.
    """
    if bars.empty:
        return pd.Series(dtype=float)
    required = {"open", "high", "low", "close", "volume"}
    missing = required - set(bars.columns)
    if missing:
        raise PineError(f"bars DataFrame missing columns: {sorted(missing)}")
    result = _eval(node, bars)
    if isinstance(result, pd.Series):
        return result
    return _as_series(result, bars.index)


def required_lookback(node: Node) -> int:
    """Return the maximum rolling window length referenced in the expression.

    Used by the engine to verify each ticker has enough history prior to the
    signal bar. Functions like crossover/crossunder add 1 bar of lookback.
    """
    max_len = 0

    def visit(n: Node) -> None:
        nonlocal max_len
        if isinstance(n, Call):
            if (
                n.name in ROLLING_FUNCS
                and len(n.args) == 2
                and isinstance(n.args[1], Num)
            ):
                max_len = max(max_len, int(n.args[1].value))
            if n.name == "atr" and len(n.args) == 1 and isinstance(n.args[0], Num):
                max_len = max(max_len, int(n.args[0].value))
            if n.name in {"crossover", "crossunder"}:
                max_len = max(max_len, 1)
            for arg in n.args:
                visit(arg)
        elif isinstance(n, (BinOp, Compare, BoolOp)):
            visit(n.left)
            visit(n.right)
        elif isinstance(n, (UnaryOp, Not)):
            visit(n.operand)

    visit(node)
    return max_len

File: screener-rs/src/pine.rs
use crate::data::Bars;
use crate::indicators;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum PineError {
    #[error("{0}")]
    Syntax(String),
    #[error("{0}")]
    Name(String),
    #[error("{0}")]
    Eval(String),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Node {
    Num(f64),
    Name(String),
    Unary {
        op: char,
        operand: Box<Node>,
    },
    Not(Box<Node>),
    Bin {
        op: char,
        left: Box<Node>,
        right: Box<Node>,
    },
    Compare {
        op: String,
        left: Box<Node>,
        right: Box<Node>,
    },
    Bool {
        op: String,
        left: Box<Node>,
        right: Box<Node>,
    },
    Call {
        name: String,
        args: Vec<Node>,
        col: usize,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum TokenKind {
    Num,
    Name,
    Op,
    Lp,
    Rp,
    Comma,
    End,
}

#[derive(Debug, Clone)]
struct Token {
    kind: TokenKind,
    value: String,
    col: usize,
}

fn tokenize(expr: &str) -> Result<Vec<Token>, PineError> {
    let chars: Vec<char> = expr.chars().collect();
    let mut out = Vec::new();
    let mut i = 0;
    while i < chars.len() {
        let ch = chars[i];
        if ch.is_whitespace() {
            i += 1;
            continue;
        }
        if ch.is_ascii_digit()
            || (ch == '.' && i + 1 < chars.len() && chars[i + 1].is_ascii_digit())
        {
            let start = i;
            let mut saw_dot = ch == '.';
            i += 1;
            while i < chars.len() && (chars[i].is_ascii_digit() || (chars[i] == '.' && !saw_dot)) {
                if chars[i] == '.' {
                    saw_dot = true;
                }
                i += 1;
            }
            out.push(Token {
                kind: TokenKind::Num,
                value: chars[start..i].iter().collect(),
                col: start,
            });
            continue;
        }
        if ch.is_ascii_alphabetic() || ch == '_' {
            let start = i;
            i += 1;
            while i < chars.len() && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') {
                i += 1;
            }
            out.push(Token {
                kind: TokenKind::Name,
                value: chars[start..i].iter().collect(),
                col: start,
            });
            continue;
        }
        if i + 1 < chars.len() {
            let two: String = chars[i..i + 2].iter().collect();
            if [">=", "<=", "==", "!="].contains(&two.as_str()) {
                out.push(Token {
                    kind: TokenKind::Op,
                    value: two,
                    col: i,
                });
                i += 2;
                continue;
            }
        }
        let kind = match ch {
            '(' => TokenKind::Lp,
            ')' => TokenKind::Rp,
            ',' => TokenKind::Comma,
            '+' | '-' | '*' | '/' | '>' | '<' => TokenKind::Op,
            _ => {
                return Err(PineError::Syntax(format!(
                    "Unexpected character {ch:?} at column {i}"
                )));
            }
        };
        out.push(Token {
            kind,
            value: ch.to_string(),
            col: i,
        });
        i += 1;
    }
    out.push(Token {
        kind: TokenKind::End,
        value: String::new(),
        col: chars.len(),
    });
    Ok(out)
}

struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn consume(&mut self) -> Token {
        let tok = self.tokens[self.pos].clone();
        self.pos += 1;
        tok
    }

    fn expect(&mut self, kind: TokenKind, value: Option<&str>) -> Result<Token, PineError> {
        let tok = self.peek();
        if tok.kind != kind || value.is_some_and(|v| tok.value != v) {
            let expected = value.unwrap_or("token");
            return Err(PineError::Syntax(format!(
                "Expected {expected:?} at column {}, got {:?}",
                tok.col, tok.value
            )));
        }
        Ok(self.consume())
    }

    fn parse(mut self) -> Result<Node, PineError> {
        let node = self.parse_or()?;
        if self.peek().kind != TokenKind::End {
            return Err(PineError::Syntax(format!(
                "Unexpected {:?} at column {}",
                self.peek().value,
                self.peek().col
            )));
        }
        Ok(node)
    }

    fn parse_or(&mut self) -> Result<Node, PineError> {
        let mut left = self.parse_and()?;
        while self.peek().kind == TokenKind::Name && self.peek().value == "or" {
            self.consume();
            let right = self.parse_and()?;
            left = Node::Bool {
                op: "or".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Node, PineError> {
        let mut left = self.parse_not()?;
        while self.peek().kind == TokenKind::Name && self.peek().value == "and" {
            self.consume();
            let right = self.parse_not()?;
            left = Node::Bool {
                op: "and".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_not(&mut self) -> Result<Node, PineError> {
        if self.peek().kind == TokenKind::Name && self.peek().value == "not" {
            self.consume();
            return Ok(Node::Not(Box::new(self.parse_not()?)));
        }
        self.parse_compare()
    }

    fn parse_compare(&mut self) -> Result<Node, PineError> {
        let left = self.parse_add()?;
        let tok = self.peek().clone();
        if tok.kind == TokenKind::Op
            && [">", ">=", "<", "<=", "==", "!="].contains(&tok.value.as_str())
        {
            self.consume();
            let right = self.parse_add()?;
            return Ok(Node::Compare {
                op: tok.value,
                left: Box::new(left),
                right: Box::new(right),
            });
        }
        Ok(left)
    }

    fn parse_add(&mut self) -> Result<Node, PineError> {
        let mut left = self.parse_mul()?;
        while self.peek().kind == TokenKind::Op && ["+", "-"].contains(&self.peek().value.as_str())
        {
            let op = self.consume().value.chars().next().unwrap();
            let right = self.parse_mul()?;
            left = Node::Bin {
                op,
                left: Box::new(left),
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_mul(&mut self) -> Result<Node, PineError> {
        let mut left = self.parse_unary()?;
        while self.peek().kind == TokenKind::Op && ["*", "/"].contains(&self.peek().value.as_str())
        {
            let op = self.consume().value.chars().next().unwrap();
            let right = self.parse_unary()?;
            left = Node::Bin {
                op,
                left: Box::new(left),
                right: Box::new(right),
            };
        }
        Ok(left)
    }

    fn parse_unary(&mut self) -> Result<Node, PineError> {
        if self.peek().kind == TokenKind::Op && ["+", "-"].contains(&self.peek().value.as_str()) {
            let op = self.consume().value.chars().next().unwrap();
            return Ok(Node::Unary {
                op,
                operand: Box::new(self.parse_unary()?),
            });
        }
        self.parse_primary()
    }

    fn parse_primary(&mut self) -> Result<Node, PineError> {
        let tok = self.peek().clone();
        match tok.kind {
            TokenKind::Num => {
                self.consume();
                Ok(Node::Num(tok.value.parse().map_err(|_| {
                    PineError::Syntax(format!(
                        "Invalid number {:?} at column {}",
                        tok.value, tok.col
                    ))
                })?))
            }
            TokenKind::Lp => {
                self.consume();
                let node = self.parse_or()?;
                self.expect(TokenKind::Rp, None)?;
                Ok(node)
            }
            TokenKind::Name => {
                self.consume();
                if self.peek().kind == TokenKind::Lp {
                    self.consume();
                    let mut args = Vec::new();
                    if self.peek().kind != TokenKind::Rp {
                        args.push(self.parse_or()?);
                        while self.peek().kind == TokenKind::Comma {
                            self.consume();
                            args.push(self.parse_or()?);
                        }
                    }
                    self.expect(TokenKind::Rp, None)?;
                    return Ok(Node::Call {
                        name: tok.value,
                        args,
                        col: tok.col,
                    });
                }
                if tok.value == "true" {
                    return Ok(Node::Num(1.0));
                }
                if tok.value == "false" {
                    return Ok(Node::Num(0.0));
                }
                Ok(Node::Name(tok.value))
            }
            _ => Err(PineError::Syntax(format!(
                "Unexpected token {:?} at column {}",
                tok.value, tok.col
            ))),
        }
    }
}

pub fn parse(expr: &str) -> Result<Node, PineError> {
    if expr.trim().is_empty() {
        return Err(PineError::Syntax("Empty expression".to_string()));
    }
    Parser {
        tokens: tokenize(expr)?,
        pos: 0,
    }
    .parse()
}

pub fn evaluate(node: &Node, bars: &Bars) -> Result<Vec<f64>, PineError> {
    if bars.is_empty() {
        return Ok(Vec::new());
    }
    eval(node, bars)
}

fn scalar(value: f64, len: usize) -> Vec<f64> {
    vec![value; len]
}

fn as_bool(value: f64) -> bool {
    value != 0.0
}

fn eval(node: &Node, bars: &Bars) -> Result<Vec<f64>, PineError> {
    let len = bars.len();
    match node {
        Node::Num(value) => Ok(scalar(*value, len)),
        Node::Name(name) => bars
            .series(name)
            .ok_or_else(|| PineError::Name(format!("Unknown identifier: {name:?}"))),
        Node::Unary { op, operand } => {
            let values = eval(operand, bars)?;
            Ok(match op {
                '-' => values.into_iter().map(|v| -v).collect(),
                '+' => values,
                _ => return Err(PineError::Syntax(format!("Unknown operator: {op:?}"))),
            })
        }
        Node::Not(operand) => Ok(eval(operand, bars)?
            .into_iter()
            .map(|v| (!as_bool(v)) as i32 as f64)
            .collect()),
        Node::Bin { op, left, right } => {
            let left = eval(left, bars)?;
            let right = eval(right, bars)?;
            Ok(left
                .iter()
                .zip(right.iter())
                .map(|(a, b)| match op {
                    '+' => a + b,
                    '-' => a - b,
                    '*' => a * b,
                    '/' => a / b,
                    _ => f64::NAN,
                })
                .collect())
        }
        Node::Compare { op, left, right } => {
            let left = eval(left, bars)?;
            let right = eval(right, bars)?;
            Ok(left
                .iter()
                .zip(right.iter())
                .map(|(a, b)| {
                    let hit = match op.as_str() {
                        ">" => a > b,
                        ">=" => a >= b,
                        "<" => a < b,
                        "<=" => a <= b,
                        "==" => a == b,
                        "!=" => a != b,
                        _ => false,
                    };
                    hit as i32 as f64
                })
                .collect())
        }
        Node::Bool { op, left, right } => {
            let left = eval(left, bars)?;
            let right = eval(right, bars)?;
            Ok(left
                .iter()
                .zip(right.iter())
                .map(|(a, b)| match op.as_str() {
                    "and" => (as_bool(*a) && as_bool(*b)) as i32 as f64,
                    "or" => (as_bool(*a) || as_bool(*b)) as i32 as f64,
                    _ => 0.0,
                })
                .collect())
        }
        Node::Call { name, args, col } => eval_call(name, args, *col, bars),
    }
}

fn int_literal(node: &Node, func: &str, arg: &str) -> Result<usize, PineError> {
    let Node::Num(value) = node else {
        return Err(PineError::Syntax(format!(
            "{func}() argument {arg:?} must be an integer literal"
        )));
    };
    let n = *value as usize;
    if *value <= 0.0 || n as f64 != *value {
        return Err(PineError::Syntax(format!(
            "{func}() argument {arg:?} must be a positive integer, got {value}"
        )));
    }
    Ok(n)
}

fn eval_call(name: &str, args: &[Node], col: usize, bars: &Bars) -> Result<Vec<f64>, PineError> {
    match name {
        "sma" | "ema" | "rsi" | "highest" | "lowest" => {
            if args.len() != 2 {
                return Err(PineError::Syntax(format!(
                    "{name}() takes 2 arguments (source, length), got {}",
                    args.len()
                )));
            }
            let source = eval(&args[0], bars)?;
            let length = int_literal(&args[1], name, "length")?;
            Ok(match name {
                "sma" => indicators::sma(&source, length),
                "ema" => indicators::ema(&source, length),
                "rsi" => indicators::rsi(&source, length),
                "highest" => indicators::highest(&source, length),
                "lowest" => indicators::lowest(&source, length),
                _ => unreachable!(),
            })
        }
        "atr" => {
            if args.len() != 1 {
                return Err(PineError::Syntax(format!(
                    "atr() takes 1 argument (length), got {}",
                    args.len()
                )));
            }
            let length = int_literal(&args[0], "atr", "length")?;
            Ok(indicators::atr(bars, length))
        }
        "crossover" | "crossunder" => {
            if args.len() != 2 {
                return Err(PineError::Syntax(format!(
                    "{name}() takes 2 arguments, got {}",
                    args.len()
                )));
            }
            let a = eval(&args[0], bars)?;
            let b = eval(&args[1], bars)?;
            Ok(if name == "crossover" {
                indicators::crossover(&a, &b)
            } else {
                indicators::crossunder(&a, &b)
            })
        }
        _ => Err(PineError::Name(format!(
            "Unknown function: {name:?} at column {col}"
        ))),
    }
}

pub fn required_lookback(node: &Node) -> usize {
    fn visit(node: &Node, max_len: &mut usize) {
        match node {
            Node::Call { name, args, .. } => {
                if ["sma", "ema", "rsi", "highest", "lowest"].contains(&name.as_str())
                    && args.len() == 2
                    && let Node::Num(n) = args[1]
                {
                    *max_len = (*max_len).max(n as usize);
                }
                if name == "atr"
                    && args.len() == 1
                    && let Node::Num(n) = args[0]
                {
                    *max_len = (*max_len).max(n as usize);
                }
                if ["crossover", "crossunder"].contains(&name.as_str()) {
                    *max_len = (*max_len).max(1);
                }
                for arg in args {
                    visit(arg, max_len);
                }
            }
            Node::Bin { left, right, .. }
            | Node::Compare { left, right, .. }
            | Node::Bool { left, right, .. } => {
                visit(left, max_len);
                visit(right, max_len);
            }
            Node::Unary { operand, .. } | Node::Not(operand) => visit(operand, max_len),
            Node::Num(_) | Node::Name(_) => {}
        }
    }
    let mut max_len = 0;
    visit(node, &mut max_len);
    max_len
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::{Bar, Bars};
    use chrono::NaiveDate;

    fn bars() -> Bars {
        let mut rows = Vec::new();
        for i in 0..5 {
            rows.push(Bar {
                date: NaiveDate::from_ymd_opt(2024, 1, 1 + i).unwrap(),
                open: 10.0 + i as f64,
                high: 11.0 + i as f64,
                low: 9.0 + i as f64,
                close: 10.0 + i as f64,
                volume: 100.0,
                adj_close: None,
                dividend: None,
            });
        }
        Bars::new(rows)
    }

    #[test]
    fn parses_and_evaluates_sma_comparison() {
        let ast = parse("close > sma(close, 3)").unwrap();
        assert_eq!(required_lookback(&ast), 3);
        let out = evaluate(&ast, &bars()).unwrap();
        assert_eq!(out, vec![0.0, 0.0, 1.0, 1.0, 1.0]);
    }
}
